### PR TITLE
Update the install generator to add Webpacker configuration, Add ReScript component and Fix Issue #1397

### PR DESCRIPTION
**Pull request by gscarv13**
### Generator update
Updated the generator to add Webpack configuration on `rails generate react_on_rails:install`
1.  Add React, Typescript, and CSS modules dependencies to the generator
2.  Add Babel config and webpacker.yml to the 
3.  Add Webpack config files

### ReScript Example
- Add ReScript example component to the `spec/dummy/`
- Boilerplate code provided by `ljuster` at shakacode/react_on_rails#1215

### Display error when SSR config is missing
- Fix issue #1397

---
This change is [Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1399)

### CODE REVIEW

File: spec/dummy/bsconfig.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  "package-specs": [
+    {
+      "module": "commonjs",
+      "in-source": true
+    }
+  ],
+  "bs-dependencies": [
+    "@rescript/react"
+  ]
+}
```
Comment (justin808) [inline spec/dummy/bsconfig.json:24]: @gscarv13 this means you're missing a return at the end of the file.

File: lib/generators/react_on_rails/templates/base/base/babel.config.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
+module.exports = function (api) {
```
Comment (justin808) [inline lib/generators/react_on_rails/templates/base/base/babel.config.js:4]: This is going to be hard to maintain.

Any thoughts on how we can rely more on rails/webpacker's defaults and then maybe modify those?

**Review by justin808 — COMMENTED** [review]
Really great work. Let's discuss this on zoom!

**Comment by justin808** [comment]
> Display error when SSR config is missing
> Fix issue If config.server_bundle_js_file is blank, nil, commented out, and prerendering is done, page request hangs #1397

Please make separate PR.

**Comment by justin808** [comment]
Any feedback on https://github.com/shakacode/react_on_rails/blob/master/CONTRIBUTING.md ?

**Comment by justin808** [comment]
`/spec/dummy` webpack config should roughly match the generator which should roughly match https://github.com/shakacode/react_on_rails_tutorial_with_ssr_and_hmr_fast_refresh

**Review by justin808 — CHANGES_REQUESTED** [review]
> Reviewed 27 of 28 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1399)* status: 27 of 28 files reviewed, 16 unresolved discussions (waiting on @gscarv13 and @justin808)

___
*[lib/generators/react_on_rails/templates/base/base/babel.config.js, line 1 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MmizP6e52MOz72kWABd-r1-1:-MnSS9zc-DJi90mpC60a:b7ujjbl) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/lib/generators/react_on_rails/templates/base/base/babel.config.js#L1)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

This is going to be hard to maintain.

Any thoughts on how we can rely more on rails/webpacker's defaults and then maybe modify those?
</blockquote></details>

Let's leave a comment at top indicating the source of this file, https://github.com/shakacode/react_on_rails_tutorial_with_ssr_and_hmr_fast_refresh
___
*[lib/generators/react_on_rails/templates/base/base/app/javascript/bundles/HelloWorld/components/HelloWorld.module.css, line 2 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSSR_CDfZ8RaxSLzzQ:-MnSSR_DEcMYFCao0OqU:bweun1y) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/lib/generators/react_on_rails/templates/base/base/app/javascript/bundles/HelloWorld/components/HelloWorld.module.css#L2)):*
> ```CSS
> .bold {
>     color: green;
> ```

bold and color green?

maybe call the style "bright"?

we can add the font-weight: bold as well
___
*[lib/generators/react_on_rails/templates/base/base/app/javascript/bundles/HelloWorld/components/HelloWorldServer.js, line 5 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSSgw00OahccR2B6zY:-MnSSgw00OahccR2B6zZ:b-mpgmyf) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/lib/generators/react_on_rails/templates/base/base/app/javascript/bundles/HelloWorld/components/HelloWorldServer.js#L5)):*
> ```JavaScript
> // For example, if using React-Router, we'd have the SSR setup here.
> 
> export default HelloWorld;
> ```

weird characters at the end of the line
___
*[lib/generators/react_on_rails/templates/base/base/config/webpacker.yml, line 1 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSSuLxCN_szGaA2tNA:-MnSSuLxCN_szGaA2tNB:bau4y7k) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/lib/generators/react_on_rails/templates/base/base/config/webpacker.yml#L1)):*
> ```YAML
> # Note: You must restart bin/webpack-dev-server for changes to take effect
> ```

leave comment regarding source came from https://github.com/shakacode/react_on_rails_tutorial_with_ssr_and_hmr_fast_refresh
___
*[lib/generators/react_on_rails/templates/base/base/config/webpack/development.js, line 1 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnST4ljETOwGy3kFkom:-MnST4ljETOwGy3kFkon:bl86103) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/lib/generators/react_on_rails/templates/base/base/config/webpack/development.js#L1)):*
> ```JavaScript
> process.env.NODE_ENV = process.env.NODE_ENV || 'development'
> ```

for all these copied files, leave comment indicating source file in https://github.com/shakacode/react_on_rails_tutorial_with_ssr_and_hmr_fast_refresh/<full-path on master>
___
*[lib/generators/react_on_rails/templates/base/base/config/webpack/serverWebpackConfig.js, line 69 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSTlda4vTdcawo4Nax:-MnSTlda4vTdcawo4Nay:b-g49679) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/lib/generators/react_on_rails/templates/base/base/config/webpack/serverWebpackConfig.js#L69)):*
> ```JavaScript
>   // the client build will handle exporting CSS.
>   // replace file-loader with null-loader
>   const rules = serverWebpackConfig.module.rules;
> ```

we can add doc resource comment:

https://webpack.js.org/configuration/module/#modulerules
___
*[lib/generators/react_on_rails/templates/base/base/config/webpack/serverWebpackConfig.js, line 105 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSTN5048ubWkPoiRUV:-MnSTN5048ubWkPoiRUW:bb69e8o) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/lib/generators/react_on_rails/templates/base/base/config/webpack/serverWebpackConfig.js#L105)):*
> ```JavaScript
>   // TODO: DELETE NEXT 2 LINES
>   // Critical due to https://github.com/rails/webpacker/pull/2644
>   // delete serverWebpackConfig.devServer
> ```

these need to be deleted ^^^^ (both repos)
___
*[lib/react_on_rails/helper.rb, line 58 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSU9lICKce1dWNY4PA:-MnSU9lICKce1dWNY4PB:b-xqb90l) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/lib/react_on_rails/helper.rb#L58)):*
> ```Ruby
>     def react_component(component_name, options = {})
>       config_server_bundle_js = ReactOnRails.configuration.server_bundle_js_file
>       if options[:prerender] == true && (config_server_bundle_js.nil? || config_server_bundle_js == "")
> ```

use `config_server_bundle.js.blank?`
___
*[lib/react_on_rails/helper.rb, line 65 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSU5zp-i_rTaOkHdtz:-MnSU5zp-i_rTaOkHdu-:bvixjay) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/lib/react_on_rails/helper.rb#L65)):*
> ```Ruby
>           Read more at https://www.shakacode.com/react-on-rails/docs/guides/react-server-rendering/
>         MSG
>         raise ReactOnRails::Error, msg
> ```

separate PR
___
*[rakelib/task_helpers.rb, line 3 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSUGre5qt-mWxoQSUn:-MnSUGre5qt-mWxoQSUo:buk1bx9) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/rakelib/task_helpers.rb#L3)):*
> ```Ruby
> # frozen_string_literal: true
> 
> require "bundler"
> ```

why was this needed?

how do you reproduce the issue that caused you to add this?
___
*[spec/dummy/bsconfig.json, line 24 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MmizP6e52MOz72kWABt-r1-24:-MnSUZms-dtSmk7WqNaI:bf3plki) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/spec/dummy/bsconfig.json#L24)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

@gscarv13 this means you're missing a return at the end of the file.
</blockquote></details>

@alexfedoseev can you confirm this is a good basic rescript configuration for React on Rails projects?
___
*[spec/dummy/Procfile.dev, line 6 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSUjoxB1IMOUF3wde9:-MnSUjoxB1IMOUF3wdeA:b-llatfy) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/spec/dummy/Procfile.dev#L6)):*
> ```
> 
> # Bundle ReScript .res files
> rescript: yarn re:start
> ```

We need to make sure that assets precompile will build.

`RAILS_ENV=production NODE_ENV=production rake assets:precompile` ==> then launch app `rails -e production`

We should add a `Procfile` with just that one line to start the rails server and comments that suggest how to build the assets.
___
*[spec/dummy/package.json, line 61 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSVDG76fhSxvRs0BLD:-MnSVDG76fhSxvRs0BLE:b-7ywkgm) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/spec/dummy/package.json#L61)):*
> ```JSON
>     "build:dev:watch": "RAILS_ENV=development NODE_ENV=development bin/webpack --watch",
>     "build:clean": "rm -rf public/webpack || true",
>     "re:build": "rescript build",
> ```

this needs to be called via assets:precompile
___
*[spec/dummy/client/app/components/HelloWorldReScript.bs.js, line 29 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSVVC4BdpPvLKc4Cjn:-MnSVVC58XqI4p2YUTUY:b-o1zia2) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/spec/dummy/client/app/components/HelloWorldReScript.bs.js#L29)):*
> ```JavaScript
> 
> exports.make = make;
> /* react Not a pure module */
> ```

Remove.

This file needs to be ignored via .gitignore.
___
*[spec/dummy/client/app/packs/client-bundle.js, line 14 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSWFK2DBCLkER03MU-:-MnSWFK2DBCLkER03MU0:b-90mg1d) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/spec/dummy/client/app/packs/client-bundle.js#L14)):*
> ```JavaScript
> import PureComponentWrappedInFunction from '../components/PureComponentWrappedInFunction';
> 
> import {make as HelloWorldReScript} from '../components/HelloWorldReScript.bs';
> ```

Add a JS component file that does this line above and exports the plain JS React component.

Then import that here.
___
*[spec/dummy/config/routes.rb, line 11 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnSWth9A7gFXd-mPt32:-MnSWth9A7gFXd-mPt33:bhawwri) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/spec/dummy/config/routes.rb#L11)):*
> ```Ruby
> 
>   get "client_side_hello_world" => "pages#client_side_hello_world"
>   get "client_side_rescript_hello_world" => "pages#client_side_rescript_hello_world"
> ```

need a unit test to check this route works.

**Comment by justin808** [comment]
1. CI needs to pass
2. Document steps from clean git clone:
  1. use the generator to build a rescript app
  2. run app in spec/dummy showing rescript component
  3. run test for new rescript component.
3. Update the basic tutorial given the updated generator https://github.com/shakacode/react_on_rails/blob/master/docs/guides/tutorial.md

**Review by justin808 — CHANGES_REQUESTED** [review]
> Reviewed 13 of 13 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/shakacode/react_on_rails/1399)* status: all files reviewed, 21 unresolved discussions (waiting on @gscarv13)

___
*[lib/generators/react_on_rails/base_generator.rb, line 40 at r2](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnY-I340L4v7s12ndFG:-MnY-I340L4v7s12ndFH:b-gxw88i) ([raw file](https://github.com/shakacode/react_on_rails/blob/81008b90dfcecf19f179be2ed4233666a5e742cd/lib/generators/react_on_rails/base_generator.rb#L40)):*
> ```Ruby
>       end
> 
>       def copy_webpack_config
> ```

Generator changes should be separate PR.
___
*[lib/generators/react_on_rails/templates/base/base/babel.config.js, line 1 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MmizP6e52MOz72kWABd-r1-1:-MnY-ZHh07kkz07hk4ZE:bb6tyiv) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/lib/generators/react_on_rails/templates/base/base/babel.config.js#L1)):*
<details><summary><i>Previously, justin808 (Justin Gordon) wrote…</i></summary><blockquote>

Let's leave a comment at top indicating the source of this file, https://github.com/shakacode/react_on_rails_tutorial_with_ssr_and_hmr_fast_refresh
</blockquote></details>

DONE
___
*[lib/generators/react_on_rails/templates/base/base/babel.config.js, line 11 at r2](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnY1OYk9RKFGFqoB3gg:-MnY1OYl8lhAHqNW5XZB:b-275s6h) ([raw file](https://github.com/shakacode/react_on_rails/blob/81008b90dfcecf19f179be2ed4233666a5e742cd/lib/generators/react_on_rails/templates/base/base/babel.config.js#L11)):*
> ```JavaScript
>   const isDevelopmentEnv = api.env('development')
>   const isProductionEnv = api.env('production')
>   const isTestEnv = api.env('test')
> ```

If changing this file, change also https://github.com/shakacode/react_on_rails_tutorial_with_ssr_and_hmr_fast_refresh/blob/master/babel.config.js ?
___
*[lib/generators/react_on_rails/templates/base/base/config/webpack/development.js, line 11 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnY1h_U3epu7TtwJVzP:-MnY1h_U3epu7TtwJVzQ:bum7sk1) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/lib/generators/react_on_rails/templates/base/base/config/webpack/development.js#L11)):*
> ```
>   const isWebpackDevServer = process.env.WEBPACK_DEV_SERVER
> ```
very important
___
*[lib/generators/react_on_rails/templates/base/base/config/webpack/development.js, line 19 at r1](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnY1jKG7dUpots9vdLd:-MnY1jKH2pKZIU_kjZLp:b-jocmo) ([raw file](https://github.com/shakacode/react_on_rails/blob/165ee675728bc475227df865a0fc2068adeee5c1/lib/generators/react_on_rails/templates/base/base/config/webpack/development.js#L19)):*
> ```JavaScript
>     const ReactRefreshWebpackPlugin = require('@pmmmwh/react-refresh-webpack-plugin')
>     clientWebpackConfig.plugins.push(
>       new ReactRefreshWebpackPlugin({
> ```

Needed for HMR
___
*[lib/generators/react_on_rails/templates/base/base/config/webpack/serverWebpackConfig.js, line 1 at r2](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnY1zvX47B7vyoEjwX5:-MnY1zvX47B7vyoEjwX6:b-6ogor9) ([raw file](https://github.com/shakacode/react_on_rails/blob/81008b90dfcecf19f179be2ed4233666a5e742cd/lib/generators/react_on_rails/templates/base/base/config/webpack/serverWebpackConfig.js#L1)):*
> ```JavaScript
> /* eslint-disable no-unused-vars */
> ```

dont' do that for the whole file, maybe a single line.
___
*[lib/generators/react_on_rails/templates/base/base/config/webpack/serverWebpackConfig.js, line 2 at r2](https://reviewable.io/reviews/shakacode/react_on_rails/1399#-MnY228O6FJCXrxDPXEP:-MnY228O6FJCXrxDPXEQ:b-8bix4p) ([raw file](https://github.com/shakacode/react_on_rails/blob/81008b90dfcecf19f179be2ed4233666a5e742cd/lib/generators/react_on_rails/templates/base/base/config/webpack/serverWebpackConfig.js#L2)):*
> ```JavaScript
> /* eslint-disable no-unused-vars */
> /* eslint-disable no-param-reassign */
> ```

don't do that

**Comment by gscarv13** [comment]
Created Separated PR for 
- Display error when SSR config is missing => #1401
- Add Rescript example => #1400

**Pull request closed without merging**